devito/data/data.py

```diff
@@ -238,8 +238,9 @@ def __getitem__(self, glb_idx, comm_type, gather_rank=None):
                     stop += sendcounts[i]
                     data_slice = recvbuf[slice(start, stop, step)]
                     shape = [r.stop-r.start for r in self._distributor.all_ranges[i]]
-                    idx = [slice(r.start, r.stop, r.step)
-                           for r in self._distributor.all_ranges[i]]
+                    idx = [slice(r.start - d.glb_min, r.stop - d.glb_min, r.step)
+                           for r, d in zip(self._distributor.all_ranges[i],
```

Comment:
Something I noticed while reviewing this PR is that both loops in lines 235 and 244 are using `i` as index.
Although this is working fine here, maybe you can change it in this PR?

Reply:
Good spot, just added